rococo/repositories/mongodb/mongodb_repository.py

```diff
@@ -0,0 +1,147 @@
+
+
+    def _process_data_before_save(self, instance: VersionedModel):
+        """Method to convert VersionedModel instance to a data dictionary that can be inserted in MongoDB"""
+        super()._process_data_before_save(instance)
+        data = instance.as_dict(convert_datetime_to_iso_string=False, convert_uuids=True)
+        for field in fields(instance):
+            if data.get(field.name) is None:
+                continue
+
+            field_value = data[field.name]
+
+            if field.metadata.get('field_type') in ['entity_id', 'uuid']:
+                if isinstance(field_value, VersionedModel):
+                    field_value = str(field_value.entity_id).replace('-', '')
+                elif isinstance(field_value, dict):
+                    field_value = str(field_value.get('entity_id')).replace('-', '')
+                elif isinstance(field_value, str):
+                    field_value = field_value.replace('-', '')
+            if isinstance(field_value, UUID):
+                field_value = str(field_value).replace('-', '')
+            if isinstance(field_value, datetime):
+                field_value = field_value.strftime('%Y-%m-%d %H:%M:%S')
+
+            data[field.name] = field_value
+        return data
+
+    def save(self, instance: VersionedModel, send_message: bool = False):
+        """Save func"""
+        data = self._process_data_before_save(instance)
+        with self.adapter:
+            inst = self.get_one(self.table_name, "entity_id", {'entity_id': data.get('entity_id')})
+            self.db[f"{self.table_name}_audit"].insert_one(inst)
+            self.db[f"{self.table_name}"].get_one_and_update(
+                {'entity_id': data.get("entity_id")},
+                {'$set': data},
+            )
+            if send_message:
+                # This assumes that the instance is now in post-saved state with all the new DB updates
+                message = json.dumps(instance.as_dict(convert_datetime_to_iso_string=True))
+                self.message_adapter.send_message(self.queue_name, message)
+        return instance
+
+    def _insert(self, data: Dict, collection_name: str):
+        self.db[collection_name].insert_one(data)
+
+    def _outdate(self, data: Dict, collection_name: str):
+        collection = self.db[collection_name]
+        return collection.find_one_and_update(
+            {'entity_id': data.get("entity_id"), 'latest': True, 'active': True},
+            {'$set': {'latest': False}},
+            upsert=False,
+            return_document=ReturnDocument.AFTER
+        )
+
+    def create(self, data: Dict, collection_name: str):
+        return self.update(data, collection_name)
+
+    def create_many(self, data: List[Dict], collection_name: str):
+        self.db[collection_name].insert_many(documents=data)
+
+    def update(self, data: Dict, collection_name: str):
+        with self.client.start_session() as session:
+            with session.start_transaction():
```

Comment:
Does usage of the distributed transactions force us to run replica sets? It used to be that way...

If so, at least the deployment scripts will have to be changed (see `mongodb` service in the `docker-compose.yml`)